### DISCUSSION TITLE
Best way to implement term_json?

### DISCUSSION THREAD

**Discussion by vasily-kartashov** [tuProlog/2p-kt#9] [Q&A]
What would be the most efficient way of implementing `term_json(?term, ?json)` predicate that covers conversion between Prolog terms and JSON strings:

```prolog
term_json(hello, '"hello"').
term_json(12, "12.3").
term_json([1, 2, 3], "[1, 2, 3]").
term_json([name("Adam"), father(null)], '{"name":"Adam","father":null}').  
```

**Comment by gciatto** [comment]
Well,
a converter among terms and json already exists, in the `serialize-core` module. It does not uses exactly your encoding but something close. Consider reading the [2p-kt presentation](https://github.com/tuProlog/2p-kt-presentation/releases/tag/1.2.0-2021-06-28T100324) for details. You could then wrap the serializer/deserializer into a primitive to call it from Prolog.

More in general, I suggest writing a converter in java/Kotlin and wrap it into a primitive.

Consider reading the presentation above to know more about primitives as well.

**Reply by vasily-kartashov** [reply]
The presentation is an intimidating read, I may have missed the Json part among 472 pages :). I'll give it another chance over the weekend.

**Reply by gciatto** [reply]
😁 Yes I belive you.

Apologies for that. 

Btw it's not meant to be read sequentially. Every module has its own section. You could just read the section about serialization, which is just a few slides describing how terms are encoded into json or yaml
